src/OpenApi/src/Services/OpenApiDocumentProvider.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.OpenApi.Writers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.AspNetCore.OpenApi;
+using Microsoft.Extensions.Options;
+using System.Linq;
+using Microsoft.OpenApi.Extensions;
+
+namespace Microsoft.Extensions.ApiDescriptions;
+
+internal class OpenApiDocumentProvider(IServiceProvider serviceProvider) : IDocumentProvider
+{
+    /// <summary>
+    /// Serializes the OpenAPI document associated with a given document name to
+    /// the provided writer.
+    /// </summary>
+    /// <param name="documentName">The name of the document to resolve.</param>
+    /// <param name="writer">A text writer associated with the document to write to.</param>
+    public async Task GenerateAsync(string documentName, TextWriter writer)
+    {
+        var targetDocumentService = serviceProvider.GetRequiredKeyedService<OpenApiDocumentService>(documentName);
+        var options = serviceProvider.GetRequiredService<IOptionsSnapshot<OpenApiOptions>>();
+        var namedOption = options.Get(documentName);
+        var document = await targetDocumentService.GetOpenApiDocumentAsync();
+        var jsonWriter = new OpenApiJsonWriter(writer);
+        document.Serialize(jsonWriter, namedOption.OpenApiVersion);
+    }
+
+    /// <summary>
+    /// Provides all document names that are currently managed in the application.
+    /// </summary>
+    public IEnumerable<string> GetDocumentNames()
```

Comment:
Just curious why an enumerable rather than materializing the query now?

Reply:
The [consumption site](https://github.com/dotnet/aspnetcore/blob/316cd2eafd939bb910142a121fc9179860bf21f7/src/Tools/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs#L243) only iterates over the list so I don't think it's necessary to materialize it into memory.